### ISSUE TITLE
Crash in tutorial level "Flags and Completion" when destroying enemy's racetrack

### ISSUE THREAD

**Issue by Quipyowert2** [Globulation2/glob2#44]
In Tutorial 4, "Flags and Completion", after destroying the enemy's Swarm and then racetrack, the game crashes with an assertion failure:
```glob2: src/Game.cpp:2289: void Game::drawMapGroundBuildings(int, int, int, int, int, int, int, int, int, Uint32, std::set<Building*>*): Assertion `building' failed.```

Globulation 2 version 214dfb22 compiled with g++ 7.5.0 in openSUSE 15.2 for WSL. X server: VcXsrv 1.20.8.1

`build/src/glob2 --version`:
```
Globulation 2 - 0.9.5.0

Compiled on Feb 17 2021 at 15:32:44

Compiled with SDL version 2.0.8
Linked with SDL version 2.0.8

Featuring :
* Map version 83
* Maps up to version 58 can still be loaded
* Network Protocol version 27
This program and all related materials are GPL, see COPYING for details.
(C) 2001-2007 Stephane Magnenat, Luc-Olivier de Charriere and other contributors.
See data/authors.txt for a full list.

Type build/src/glob2 --help for a list of command line options.

```

Reproducibility:
Crashes every time when following the steps below.

Steps to reproduce:
1. Press space until the flags and zones tab appears
2. Place exploration flag near enemy base
3. Place defense zone behind your defense towers (the large square blue zone)
4. Place war flag on enemy's swarm; set radius to 0.
5. Increase number of workers supplying the swarm to about 10.
6. Increase warrior bar on your swarm to about 3 or 4.
7. When warriors destroy enemy's swarm, move war flag onto racetrack just south of where the destroyed swarm was.

Actual result:
Glob2 crashes with an assertion failure when the racetrack is destroyed.

Expected result:
Tutorial finished, maybe?

Backtrace:
```
#0  0x00007ffffc9c9520 in raise () from /lib64/libc.so.6
#1  0x00007ffffc9cab01 in abort () from /lib64/libc.so.6
#2  0x00007ffffc9c1b1a in __assert_fail_base () from /lib64/libc.so.6
#3  0x00007ffffc9c1b92 in __assert_fail () from /lib64/libc.so.6
#4  0x00000000004ef426 in Game::drawMapGroundBuildings (this=0x7ffffffe3650, left=0, top=0, right=32, bot=24, sw=1024,
    sh=768, viewportX=36, viewportY=27, localTeam=0, drawOptions=11, visibleBuildings=0x7ffffffe2f00)
    at src/Game.cpp:2289
#5  0x00000000004eba42 in Game::drawMap (this=0x7ffffffe3650, sx=0, sy=0, sw=1024, sh=768, rightMargin=160,
    topMargin=16, viewportX=36, viewportY=27, localTeam=0, drawOptions=11, visibleBuildings=0x7ffffffe2f00)
    at src/Game.cpp:2822
#6  0x000000000051854f in GameGUI::drawAll (this=0x7ffffffe35d0, team=0) at src/GameGUI.cpp:4399
#7  0x00000000004d43f0 in Engine::run (this=0x7ffffffe35d0) at src/Engine.cpp:511
#8  0x00000000004bb4df in CampaignMenuScreen::onAction (this=0x7ffffffe89c0, source=0xe0a030,
    action=GAGGUI::BUTTON_RELEASED, par1=1, par2=0) at src/CampaignMenuScreen.cpp:70
#9  0x0000000000756cdf in GAGGUI::Button::onSDLMouseButtonUp (this=0xe0a030, event=0x7ffffffe88b0)
    at libgag/src/GUIButton.cpp:84
#10 0x0000000000754a32 in GAGGUI::Screen::dispatchEvents (this=0x7ffffffe89c0, event=0x7ffffffe88b0)
    at libgag/src/GUIBase.cpp:601
#11 0x00000000007544c2 in GAGGUI::Screen::execute (this=0x7ffffffe89c0, gfx=0xc47560, stepLength=40)
    at libgag/src/GUIBase.cpp:511
#12 0x0000000000544bd2 in Glob2::run (this=0x7ffffffedb4f, argc=2, argv=0x7ffffffedc38) at src/Glob2.cpp:308
#13 0x00000000005451aa in main (argc=2, argv=0x7ffffffedc38) at src/Glob2.cpp:439 
```

**Comment by Quipyowert2** [comment]
I think this is the same bug as [savannah bug #55569](https://savannah.nongnu.org/bugs/?55569).

**Comment by Quipyowert2** [comment]
At [Building.cpp:304](https://github.com/Globulation2/glob2/blob/9a71aa19d0d7f37bd7d887d3613a339ac3da8d58/src/Building.cpp#L304), when loading `campaigns/tutorial-part4.map`, shortTypeNum is 3, which indicates a racetrack. `type->width` and ->height are both 6, but the building is only level 1, so the width and height should be only 4x4 according to `data/buildings.txt`. When I destroy the enemy racetrack with my warriors, the 4x4 area in the center of the 6x6 area where the building was is cleared, but the edge of the 6x6 area is not cleared, leaving a invalid GBID on the ground.

Later, when Globulation 2 tries to draw a building in [Game::drawMapGroundBuildings()](https://github.com/Globulation2/glob2/blob/9a71aa19d0d7f37bd7d887d3613a339ac3da8d58/src/Game.cpp#L2289), there is an attempt to convert the garbage building ID on the ground to a Building pointer, which fails and returns NULL because the building no longer exists. Then, either an assertion fails because building is NULL (in debug builds) or a segmentation fault/access violation happens when trying to dereference building.

Other places it crashes include [Unit.cpp:504](https://github.com/Globulation2/glob2/blob/9a71aa19d0d7f37bd7d887d3613a339ac3da8d58/src/Unit.cpp#L504) and [Minimap.cpp:338](https://github.com/Globulation2/glob2/blob/9a71aa19d0d7f37bd7d887d3613a339ac3da8d58/src/Minimap.cpp#L338).

**Comment by stephanemagnenat** [comment]
Thank you very much for the investigation! So it seems the call to `setBuilding` passe incorrect coordinates in some case. It seems to me that this happens when a building has an original size that is small and whose upgrades make it bigger. I suspect that somehow the original size is wrongly used fol clearing the map upon destruction.

**Comment by Quipyowert2** [comment]
Maybe calling Game::integrity or Team:integrity at certain places in the code would help identify where those bad coordinates are coming from. I forgot to mention, when my globs attack Blue's.ravetrack, they do the attack animation when there's a one tile of empty space between them and the racetrack, and it looks like it's level 1. They should be right next to the racetrack when they're attacking, right?

**Comment by stephanemagnenat** [comment]
Yes they should be next to the racetrack. So it seems that a racetrack of level 1 is having the footprint of a racetrack of level 2. I am actually wondering whether the map that we load is itself corrupted. We could probably investigate that by doing some integrity check as you suggest and making sure that every map cell referring to a building is actually below this building.

**Comment by stephanemagnenat** [comment]
I investigated by improving the integrity check in #84. Assuming I did not make a mistake there, what this illustrates is that the 4th map of the tutorial is corrupted: the building has a 6x6 footprint on the map but it is actually 4x4. I believe that it is an old bug that made its way within the map.

We could fix this by having a heal pass on map load, or by creating an additional CLI tool to do the healing.
